src/TLuaInterpreter.cpp

```diff
@@ -314,6 +314,17 @@ int TLuaInterpreter::raiseEvent( lua_State * L )
             pE.mArgumentList.append( QString::number(lua_toboolean( L, i )) );
             pE.mArgumentTypeList.append( ARGUMENT_TYPE_BOOLEAN );
         }
+        else if( lua_isnil( L, i ) )
+        {
+            pE.mArgumentList.append( QString() );
+            pE.mArgumentTypeList.append( ARGUMENT_TYPE_NIL );
+        }
+        else
+        {
+            lua_pushstring( L, "raiseEvent: invalid argument type" );
```

Comment:
I did summarise the options in a [commit](https://github.com/SlySven/Mudlet/commit/8456c13f69871e104bf5c7c7f9892eaa5625bf16) of mine. In summery:
>The current methodology is:
* For bad argument TYPE situations: to throw a lua_error and provide a
error message using a fairly uniform construction, I provide these via
Qt's translation system to permit that part of the Lua API to be
translated with the C++ parts of the application's framework.
* For bad argument VALUE situations best practice is for a lua function to
provide a primary nil return value but to also provide a secondary value as
a string message describing the (first) argument value error.  A lua error
is NOT thrown but unless the user catches the error message by providing a
second variable to contain the second return value they may not find out
what is wrong.  Again I choose to provide a translatable message for this.
* For good arguments case that are satisfactory but produce no data, an
empty table, an empty string or a signal number value is an appropriate
primary return value. For example for either type of table that the lua
getAreaExits((int) areaId, <(bool) provide Full data> ) function now
returns, an empty table will be returned for an area which is isolated and
has no exits.  A secondary return value is NOT expected in this situation
as far as I can tell, though I believe one could be useful...
* For good arguments values for functions that provide data then the data
is returned as the primary and only return value.
* For good arguments values for functions that do not return data then
success should be indicated by a single true Boolean return value.

For some examples, including using QT's GUI translation system to allow for future internationalisation (Mudlet 4.0 ? :pray:) - this replaces using `lua_pushfstring(...)` as Vadim indicates in his example, consider these `diff` file code snippets:

<pre>@@ -4184,7 +4184,8 @@ int TLuaInterpreter::getAreaExits( lua_State *L )
     TArea * pA = pHost->mpMap->mpRoomDB->getArea( area );
     if( !pA ) {
         lua_pushnil(L);
-        return 1;
+        lua_pushstring( L, tr("getAreaExits: Bad value #1 (number %1 is not a valid area ID).").arg(area).toUtf8().constData());
+        return 2;
     }</pre>
 
<pre>@@ -8438,14 +8439,35 @@ int TLuaInterpreter::setRoomArea( lua_State * L )
     }
 
     Host * pHost = TLuaInterpreter::luaInterpreterMap[L];
-    if ( area <= 0 ) {
+    if( ! pHost ) {
+        lua_pushnil( L );
+        lua_pushstring( L, tr( "setRoomArea: NULL Host pointer - something is wrong!" ).toUtf8().constData() );
+        return 2;
+    }
+    else if( ! pHost->mpMap || ! pHost->mpMap->mpRoomDB ) {
+        lua_pushnil( L );
+        lua_pushstring( L, tr( "setRoomArea: No map present or loaded!" ).toUtf8().constData() );
+        return 2;
+    }
+    else if( area <= 0 ) {
+        lua_pushnil( L );
         lua_pushstring( L, tr( "setRoomArea: bad argument #2 value (area Id must be > 0.  To remove a room's area, use resetRoomArea(roomID) )." ).toUtf8().constData() );
-        lua_error( L );
+        return 2;
+    }
+    else if( ! pHost->mpMap->mpRoomDB->getRoomIDList().contains( id ) ) {
+        lua_pushnil( L );
+        lua_pushstring( L, tr( "setRoomArea: bad argument #1 value (room Id must exist.)" ).toUtf8().constData() );
+        return 2;
+    }
+    else if( ! pHost->mpMap->mpRoomDB->getAreaIDList().contains( area ) ) {
+        lua_pushnil( L );
+        lua_pushstring( L, tr( "setRoomArea: bad argument #2 value (area Id must exist.)" ).toUtf8().constData() );
+        return 2;
+    }
+    else {
+        lua_pushboolean( L, pHost->mpMap->setRoomArea( id, area, false ) );
         return 1;
     }
-
-    lua_pushboolean( L, pHost->mpMap->setRoomArea( id, area, false ) );
-    return 1;
 }</pre>

For `QStrings` that are not to be put through translation use a `QStringLiteral("`...`")`<sup>For Qt5.0 or later versions</sup> wrapper rather than just `"`...`"` when defining literals - the means of use in code is the same as the `tr("`...`")` method call BTW.  This marks them as NOT to be translated - and protects against the compile errors that will be thrown up when the `QT_NO_CAST_FROM_ASCII` and `QT_NO_CAST_TO_ASCII` symbols are `#define`d to disable automatic conversions between "8-bit strings" `(char *)` and "unicode QStrings".  This will be done to both find strings that are not prepared for GUI translation or locked from it; and to enforce against them occurring without notice.